src/librustc_mir/transform/const_prop.rs

```diff
@@ -471,60 +463,64 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
-                        ClearCrossCrate::Clear => return None,
-                    };
-                    let dir = if *op == BinOp::Shr {
-                        "right"
-                    } else {
-                        "left"
-                    };
-                    let hir_id = source_scope_local_data[source_info.scope].lint_root;
-                    self.tcx.lint_hir(
-                        ::rustc::lint::builtin::EXCEEDING_BITSHIFTS,
-                        hir_id,
-                        span,
-                        &format!("attempt to shift {} with overflow", dir));
-                    return None;
+                // If overflow checking is enabled (like in debug mode by default),
+                // then we'll already catch overflow when we evaluate the `Assert` statement
+                // in MIR. However, if overflow checking is disabled, then there won't be any
+                // `Assert` statement and so we have to do additional checking here.
+                if !overflow_check {
+                    self.use_ecx(source_info, |this| {
+                        let l = this.ecx.read_immediate(this.ecx.eval_operand(left, None)?)?;
+                        let (_, overflow, _ty) = this.ecx.overflowing_binary_op(*op, l, r)?;
+
+                        if overflow {
+                            let err = err_panic!(Overflow(*op)).into();
+                            return Err(err);
+                        }
+
+                        Ok(())
+                    })?;
                 }
             }
-            self.use_ecx(source_info, |this| {
-                let l = this.ecx.read_immediate(this.ecx.eval_operand(left, None)?)?;
-                let (_, overflow, _ty) = this.ecx.overflowing_binary_op(*op, l, r)?;
-
-                // We check overflow in debug mode already
-                // so should only check in release mode.
-                if !this.tcx.sess.overflow_checks() && overflow {
-                    let err = err_panic!(Overflow(*op)).into();
-                    return Err(err);
-                }
 
-                Ok(())
-            })?;
-        } else if let Rvalue::Ref(_, _, place) = rvalue {
-            trace!("checking Ref({:?})", place);
-            // FIXME(wesleywiser) we don't currently handle the case where we try to make a ref
-            // from a function argument that hasn't been assigned to in this function.
-            if let Place {
-                base: PlaceBase::Local(local),
-                projection: box []
-            } = place {
+            Rvalue::Ref(_, _, Place { base: PlaceBase::Local(local), projection: box [] }) => {
+                trace!("checking Ref({:?})", place);
+                // FIXME(wesleywiser) we don't currently handle the case where we try to make a ref
+                // from a function argument that hasn't been assigned to in this function.
                 let alive =
```

Comment:
Why is this still needed, with the new machine hook and the const_prop machine?

Reply:
The alloc_local hook runs too late to fix this ICE because miri calls `InterpCx::size_and_align_of()` before accessing the local which expects `metadata` to have a value. So this is still necessary.

Reply:
Does that mean `access_local` is not actually helping?
Is there a way we could reorder things to avoid this? (I have no idea which call to `size_and_align_of` is the problematic one.)

Either way, this should be explained in a comment the code, not just in the PR.

Reply:
The `access_local` hook works around a different issue which is that we may try to load values from uninitialized locals. In const prop, this happens for a few different reasons but mostly related to this block here

https://github.com/rust-lang/rust/blob/53aca553ec08e62a0591806fcf0f706ee9f53778/src/librustc_mir/transform/const_prop.rs#L625-L645

If we aren't allowed to const prop a local, it will `LocalValue::Uninitialized` when `ConstPropMachine` tries to access it.

Reply:
I see. So which call to `size_and_align_of` is the one that is guarded by the check here?

Reply:
It's this call here:

https://github.com/rust-lang/rust/blob/a16dca337de610986252bb800953e57bf395863f/src/librustc_mir/interpret/place.rs#L986


Full stacktrace for the ICE you get if you remove this check:

<details>

```
wesley@endurance:~/code/rust/rust> RUST_BACKTRACE=1 rustc +stage1 -Zmir-opt-level=3 src/test/ui/unsized-locals/autoderef.rs 
thread 'rustc' panicked at 'slice fat ptr must have vtable', src/libcore/option.rs:1186:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /home/wesley/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /home/wesley/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1028
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:189
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:206
  10: <alloc::boxed::Box<F> as core::ops::function::Fn<A>>::call
             at ./src/liballoc/boxed.rs:946
  11: rustc_driver::report_ice
             at src/librustc_driver/lib.rs:1187
  12: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:473
  13: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:376
  14: rust_begin_unwind
             at src/libstd/panicking.rs:303
  15: core::panicking::panic_fmt
             at src/libcore/panicking.rs:84
  16: core::option::expect_failed
             at src/libcore/option.rs:1186
  17: core::option::Option<T>::expect
             at ./src/libcore/option.rs:345
  18: rustc_mir::interpret::eval_context::InterpCx<M>::size_and_align_of
             at src/librustc_mir/interpret/eval_context.rs:460
  19: rustc_mir::interpret::place::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::force_allocation_maybe_sized
             at src/librustc_mir/interpret/place.rs:989
  20: rustc_mir::interpret::place::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::force_allocation
             at src/librustc_mir/interpret/place.rs:1020
  21: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_rvalue_into_place
             at src/librustc_mir/interpret/step.rs:244
  22: rustc_mir::transform::const_prop::ConstPropagator::const_prop::{{closure}}
             at src/librustc_mir/transform/const_prop.rs:546
  23: rustc_mir::transform::const_prop::ConstPropagator::use_ecx
             at src/librustc_mir/transform/const_prop.rs:367
  24: rustc_mir::transform::const_prop::ConstPropagator::const_prop
             at src/librustc_mir/transform/const_prop.rs:544
  25: <rustc_mir::transform::const_prop::ConstPropagator as rustc::mir::visit::MutVisitor>::visit_statement
             at src/librustc_mir/transform/const_prop.rs:711
  26: rustc::mir::visit::MutVisitor::super_basic_block_data
             at ./src/librustc/mir/visit.rs:312
  27: rustc::mir::visit::MutVisitor::visit_basic_block_data
             at ./src/librustc/mir/visit.rs:81
  28: rustc::mir::visit::MutVisitor::super_body
             at ./src/librustc/mir/visit.rs:270
  29: rustc::mir::visit::MutVisitor::visit_body
             at ./src/librustc/mir/visit.rs:75
  30: <rustc_mir::transform::const_prop::ConstProp as rustc_mir::transform::MirPass>::run_pass
             at src/librustc_mir/transform/const_prop.rs:105
  31: rustc_mir::transform::run_passes::{{closure}}
             at src/librustc_mir/transform/mod.rs:174
  32: rustc_mir::transform::run_passes
             at src/librustc_mir/transform/mod.rs:181
  33: rustc_mir::transform::run_optimization_passes
             at src/librustc_mir/transform/mod.rs:229
  34: rustc_mir::transform::optimized_mir
             at src/librustc_mir/transform/mod.rs:297
  35: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::optimized_mir>::compute
             at src/librustc/ty/query/plumbing.rs:934
  36: rustc::dep_graph::graph::DepGraph::with_task_impl
             at src/librustc/dep_graph/graph.rs:321
  37: rustc::dep_graph::graph::DepGraph::with_task
             at src/librustc/dep_graph/graph.rs:197
  38: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::force_query_with_job::{{closure}}::{{closure}}
             at src/librustc/ty/query/plumbing.rs:516
  39: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::start_query::{{closure}}::{{closure}}
             at src/librustc/ty/query/plumbing.rs:248
  40: rustc::ty::context::tls::enter_context::{{closure}}
             at src/librustc/ty/context.rs:1864
  41: rustc::ty::context::tls::set_tlv
             at src/librustc/ty/context.rs:1797
  42: rustc::ty::context::tls::enter_context
             at src/librustc/ty/context.rs:1863
  43: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::start_query::{{closure}}
             at src/librustc/ty/query/plumbing.rs:247
  44: rustc::ty::context::tls::with_related_context::{{closure}}
             at src/librustc/ty/context.rs:1970
  45: rustc::ty::context::tls::with_context::{{closure}}
             at src/librustc/ty/context.rs:1953
  46: rustc::ty::context::tls::with_context_opt
             at src/librustc/ty/context.rs:1942
  47: rustc::ty::context::tls::with_context
             at src/librustc/ty/context.rs:1953
  48: rustc::ty::context::tls::with_related_context
             at src/librustc/ty/context.rs:1966
  49: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::start_query
             at src/librustc/ty/query/plumbing.rs:236
  50: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::force_query_with_job::{{closure}}
             at src/librustc/ty/query/plumbing.rs:508
  51: rustc::ty::query::plumbing::with_diagnostics
             at src/librustc/ty/query/plumbing.rs:181
  52: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::force_query_with_job
             at src/librustc/ty/query/plumbing.rs:507
  53: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
             at src/librustc/ty/query/plumbing.rs:344
  54: rustc_mir::monomorphize::collector::collect_neighbours
             at src/librustc_mir/monomorphize/collector.rs:1250
  55: rustc_mir::monomorphize::collector::collect_items_rec
             at src/librustc_mir/monomorphize/collector.rs:403
  56: rustc_mir::monomorphize::collector::collect_items_rec
             at src/librustc_mir/monomorphize/collector.rs:413
  57: rustc_mir::monomorphize::collector::collect_crate_mono_items::{{closure}}::{{closure}}
             at src/librustc_mir/monomorphize/collector.rs:311
  58: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}
             at ./src/libcore/iter/traits/iterator.rs:613
  59: core::iter::traits::iterator::Iterator::fold::ok::{{closure}}
             at ./src/libcore/iter/traits/iterator.rs:1813
  60: core::iter::traits::iterator::Iterator::try_fold
             at ./src/libcore/iter/traits/iterator.rs:1694
  61: core::iter::traits::iterator::Iterator::fold
             at ./src/libcore/iter/traits/iterator.rs:1816
  62: core::iter::traits::iterator::Iterator::for_each
             at ./src/libcore/iter/traits/iterator.rs:616
  63: rustc_mir::monomorphize::collector::collect_crate_mono_items::{{closure}}
             at src/librustc_mir/monomorphize/collector.rs:309
  64: rustc::util::common::time_ext
             at ./src/librustc/util/common.rs:55
  65: rustc::util::common::time
             at ./src/librustc/util/common.rs:49
  66: rustc_mir::monomorphize::collector::collect_crate_mono_items
             at src/librustc_mir/monomorphize/collector.rs:308
  67: rustc_mir::monomorphize::partitioning::collect_and_partition_mono_items::{{closure}}
             at src/librustc_mir/monomorphize/partitioning.rs:890
  68: rustc::util::common::time_ext
             at ./src/librustc/util/common.rs:55
  69: rustc::util::common::time
             at ./src/librustc/util/common.rs:49
  70: rustc_mir::monomorphize::partitioning::collect_and_partition_mono_items
             at src/librustc_mir/monomorphize/partitioning.rs:889
  71: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::collect_and_partition_mono_items>::compute::{{closure}}
             at ./src/librustc/ty/query/plumbing.rs:942
  72: rustc::ty::query::__query_compute::collect_and_partition_mono_items
             at ./src/librustc/ty/query/plumbing.rs:893
  73: rustc::dep_graph::graph::DepGraph::with_task_impl
             at ./src/librustc/dep_graph/graph.rs:321
  74: rustc::dep_graph::graph::DepGraph::with_eval_always_task
             at ./src/librustc/dep_graph/graph.rs:374
  75: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::force_query_with_job::{{closure}}::{{closure}}
             at ./src/librustc/ty/query/plumbing.rs:510
  76: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::start_query::{{closure}}::{{closure}}
             at ./src/librustc/ty/query/plumbing.rs:248
  77: rustc::ty::context::tls::enter_context::{{closure}}
             at ./src/librustc/ty/context.rs:1864
  78: rustc::ty::context::tls::set_tlv
             at ./src/librustc/ty/context.rs:1797
  79: rustc::ty::context::tls::enter_context
             at ./src/librustc/ty/context.rs:1863
  80: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::start_query::{{closure}}
             at ./src/librustc/ty/query/plumbing.rs:247
  81: rustc::ty::context::tls::with_related_context::{{closure}}
             at ./src/librustc/ty/context.rs:1970
  82: rustc::ty::context::tls::with_context::{{closure}}
             at ./src/librustc/ty/context.rs:1953
  83: rustc::ty::context::tls::with_context_opt
             at ./src/librustc/ty/context.rs:1942
  84: rustc::ty::context::tls::with_context
             at ./src/librustc/ty/context.rs:1953
  85: rustc::ty::context::tls::with_related_context
             at ./src/librustc/ty/context.rs:1966
  86: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::start_query
             at ./src/librustc/ty/query/plumbing.rs:236
  87: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::force_query_with_job::{{closure}}
             at ./src/librustc/ty/query/plumbing.rs:508
  88: rustc::ty::query::plumbing::with_diagnostics
             at ./src/librustc/ty/query/plumbing.rs:181
  89: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::force_query_with_job
             at ./src/librustc/ty/query/plumbing.rs:507
  90: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
             at ./src/librustc/ty/query/plumbing.rs:344
  91: rustc::ty::query::TyCtxtAt::collect_and_partition_mono_items
             at ./src/librustc/ty/query/plumbing.rs:1019
  92: rustc::ty::query::<impl rustc::ty::context::TyCtxt>::collect_and_partition_mono_items
             at ./src/librustc/ty/query/plumbing.rs:1011
  93: rustc_codegen_ssa::base::codegen_crate
             at ./src/librustc_codegen_ssa/base.rs:504
  94: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::codegen_crate
             at src/librustc_codegen_llvm/lib.rs:293
  95: rustc_interface::passes::start_codegen::{{closure}}
             at src/librustc_interface/passes.rs:1075
  96: rustc::util::common::time_ext
             at ./src/librustc/util/common.rs:55
  97: rustc::util::common::time
             at ./src/librustc/util/common.rs:49
  98: rustc_interface::passes::start_codegen
             at src/librustc_interface/passes.rs:1073
  99: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::ongoing_codegen::{{closure}}::{{closure}}
             at src/librustc_interface/queries.rs:237
 100: rustc_interface::passes::BoxedGlobalCtxt::enter::{{closure}}::{{closure}}
             at src/librustc_interface/passes.rs:809
 101: rustc::ty::context::tls::enter_global::{{closure}}
             at ./src/librustc/ty/context.rs:1896
 102: rustc::ty::context::tls::enter_context::{{closure}}
             at ./src/librustc/ty/context.rs:1864
 103: rustc::ty::context::tls::set_tlv
             at ./src/librustc/ty/context.rs:1797
 104: rustc::ty::context::tls::enter_context
             at ./src/librustc/ty/context.rs:1863
 105: rustc::ty::context::tls::enter_global
             at ./src/librustc/ty/context.rs:1895
 106: rustc_interface::passes::BoxedGlobalCtxt::enter::{{closure}}
             at src/librustc_interface/passes.rs:809
 107: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
             at ./<::rustc_data_structures::box_region::declare_box_region_type macros>:21
 108: rustc_interface::passes::create_global_ctxt::{{closure}}
             at src/librustc_interface/passes.rs:873
 109: alloc::boxed::<impl core::ops::generator::Generator for core::pin::Pin<alloc::boxed::Box<G>>>::resume
             at ./src/liballoc/boxed.rs:1083
 110: rustc_data_structures::box_region::PinnedGenerator<I,A,R>::access
             at ./src/librustc_data_structures/box_region.rs:52
 111: rustc_interface::passes::BoxedGlobalCtxt::access
             at ./<::rustc_data_structures::box_region::declare_box_region_type macros>:24
 112: rustc_interface::passes::BoxedGlobalCtxt::enter
             at src/librustc_interface/passes.rs:809
 113: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::ongoing_codegen::{{closure}}
             at src/librustc_interface/queries.rs:231
 114: rustc_interface::queries::Query<T>::compute
             at src/librustc_interface/queries.rs:27
 115: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::ongoing_codegen
             at src/librustc_interface/queries.rs:229
 116: rustc_driver::run_compiler::{{closure}}
             at src/librustc_driver/lib.rs:387
 117: rustc_interface::interface::run_compiler_in_existing_thread_pool
             at ./src/librustc_interface/interface.rs:117
 118: rustc_interface::interface::run_compiler::{{closure}}
             at ./src/librustc_interface/interface.rs:130
 119: rustc_interface::util::spawn_thread_pool::{{closure}}::{{closure}}::{{closure}}::{{closure}}
             at ./src/librustc_interface/util.rs:192
 120: rustc::ty::context::tls::with_thread_locals::{{closure}}::{{closure}}
             at ./src/librustc/ty/context.rs:1852
 121: std::thread::local::LocalKey<T>::try_with
             at ./src/libstd/thread/local.rs:262
 122: std::thread::local::LocalKey<T>::with
             at ./src/libstd/thread/local.rs:239
 123: rustc::ty::context::tls::with_thread_locals::{{closure}}
             at ./src/librustc/ty/context.rs:1844
 124: std::thread::local::LocalKey<T>::try_with
             at ./src/libstd/thread/local.rs:262
 125: std::thread::local::LocalKey<T>::with
             at ./src/libstd/thread/local.rs:239
 126: rustc::ty::context::tls::with_thread_locals
             at ./src/librustc/ty/context.rs:1836
 127: rustc_interface::util::spawn_thread_pool::{{closure}}::{{closure}}::{{closure}}
             at ./src/librustc_interface/util.rs:192
 128: scoped_tls::ScopedKey<T>::set
             at /home/wesley/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
 129: rustc_interface::util::spawn_thread_pool::{{closure}}::{{closure}}
             at ./src/librustc_interface/util.rs:188
 130: scoped_tls::ScopedKey<T>::set
             at /home/wesley/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
 131: syntax::with_globals::{{closure}}
             at ./src/libsyntax/lib.rs:108
 132: scoped_tls::ScopedKey<T>::set
             at /home/wesley/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
 133: syntax::with_globals
             at ./src/libsyntax/lib.rs:107
 134: rustc_interface::util::spawn_thread_pool::{{closure}}
             at ./src/librustc_interface/util.rs:187
 135: rustc_interface::util::scoped_thread::{{closure}}
             at ./src/librustc_interface/util.rs:164
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.40.0-dev running on x86_64-unknown-linux-gnu

note: compiler flags: -Z mir-opt-level=3

query stack during panic:
#0 [optimized_mir] processing `<[char] as Foo>::foo`
#1 [collect_and_partition_mono_items] collect_and_partition_mono_items
end of query stack
```

</details>

Reply:
So that seems to come via this call:

https://github.com/rust-lang/rust/blob/ea45150837990ccf40edfe2725ac492fc35193a1/src/librustc_mir/interpret/step.rs#L244

That's odd though... we call `eval_place` before that so if it was a local, `access_local` could actually throw an error to never reach the later part?

Reply:
That seems to be because `eval_place()` doesn't call `access_local()`. `access_local()` returns an `Operand` which `eval_place()` wouldn't use at all since it just returns a `PlaceTy`. Maybe we could generalize `access_local()` to just do validation. Perhaps something like:

```rust
fn validate_local_access(ecx: InterpCx, frame: Frame, local: Local) -> InterpResult<'tcx>
```

Reply:
Ah, we even have a comment there saying
```
// This works even for dead/uninitialized locals; we check further when writing
```
Except for unsized locals it doesn't... or well it does but only to *initialize* them; in this stacktrace we are taking a reference to them.

> `validate_local_access`

"validate" in Miri already refers to [this](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/validity.rs) so let's avoid that term. But also, I don't think that would be right: when calling `eval_place` for the `dest`, we are fine with uninitialized locals as we will write to them!

What basically happens here is that const_prop causes the engine to take a reference to an uninitialized unsized value. That's impossible to implement (taking a reference needs an allocation in memory but to have one we need a size, which we cannot determine for an unsized uninitialized value). Maybe this check is indeed better than the possible alternatives (though it would be nice for the comment here to explain that).

The comment here talks about arguments that are fat pointers; is that correct? Fat pointers are also just sized data (twice as big as normal pointers) and shouldn't be a problem. Unsized arguments however are a problem, if my analysis is correct.

Reply:
> "validate" in Miri already refers to this so let's avoid that term.

Got it, thanks!

> The comment here talks about arguments that are fat pointers; is that correct? Fat pointers are also just sized data (twice as big as normal pointers) and shouldn't be a problem. Unsized arguments however are a problem, if my analysis is correct.

Your analysis is correct; it was early and I hadn't yet had coffee when I wrote "fat pointers". The ICE I linked to occurs when processing:

```
#0 [optimized_mir] processing `<[char] as Foo>::foo`
```

and specifically with the unsized `self` parameter.

Reply:
Good, I think we reached a conclusion here then. :) Would be good to have that documented in the code, but right now I don't see a good way to handle this inside the engine.

I am really not happy with out treatment of unsized locals in general, and this puts the finger on what doesn't work well...